### ISSUE TITLE
Can't get to work with BeforeSuite and AfterSuite hooks

### ISSUE THREAD

**Issue by AdamOakman** [codeceptjs/CodeceptJS#1430]
#### What are you trying to achieve?

Get to work with BeforeSuite and AfterSuite hooks. Simple I.amOnPage('/'); step returns errors in console.

#### What do you get instead?

> Provide console output if related. Use `--verbose` mode for more details.

```bash
CodeceptJS v2.0.1
Using test root "/tests"
Helpers: Protractor
Plugins: screenshotOnFail, allure, autoDelay

TS_005 - Konfigurator - Dodawanie hosta --
    [1] Starting recording promises
    Emitted | suite.before ([object Object])
    Emitted | hook.start ([object Object])
    Emitted | step.before (Ja jestem na_stronie "/")
    Emitted | step.after (Ja jestem na_stronie "/")
    Emitted | step.start (Ja jestem na_stronie "/")
    Context: BeforeSuite
      Ja jestem na_stronie "/"
    [1] Error | TypeError: Cannot read property 'get' of undefined
    Emitted | step.failed (Ja jestem na_stronie "/")
    Emitted | step.finish (Ja jestem na_stronie "/")
    [1] Error | TypeError: Cannot read property 'get' of undefined
    [1] Starting <teardown> session
    Emitted | test.failed ([object Object])
    Error processing test.failed event:
    TypeError: Cannot read property 'end' of undefined
    at Allure.endCase (/codecept/node_modules/allure-js-commons/index.js:50:26)
    at EventEmitter.event.dispatcher.on (/codecept/lib/plugin/allure.js:91:14)
    at EventEmitter.emit (events.js:187:15)
    at Object.emit (/codecept/lib/event.js:49:28)
    at recorder.errHandler (/codecept/lib/scenario.js:99:13)
    at promise.catch (/codecept/lib/recorder.js:173:9)
    at process._tickCallback (internal/process/next_tick.js:68:7)
    [1] <teardown> Stopping recording promises
 › <screenshotOnFail> Test failed, saving screenshot
  ✖ FAILED in 7ms

    Emitted | hook.start ([object Object])
    [2] Starting recording promises
    Emitted | step.before (Ja jestem na_stronie "/")
    Emitted | step.after (Ja jestem na_stronie "/")
    Emitted | step.start (Ja jestem na_stronie "/")
    Context: AfterSuite
      Ja jestem na_stronie "/"
    [2] Error | TypeError: Cannot read property 'get' of undefined
    Emitted | step.failed (Ja jestem na_stronie "/")
    Emitted | step.finish (Ja jestem na_stronie "/")
    [2] Error | TypeError: Cannot read property 'get' of undefined
    [2] Stopping recording promises
```

> Provide test source code if related

```js
Feature('TS_005 - Konfigurator - Dodawanie hosta');

BeforeSuite( (I) => {
	I.amOnPage('/');
	// I.amOnConfigMainPage();
	// I.loginAs('operator');
	// I.addNewClient(client);
});

Before((I) => {
  	I.amOnConfigMainPage();
	I.loginAs('operator');
});

Data(testdata).Scenario('TS_005 - TC_001 - Dodaj hosta', (I, current) => {

	I.addNewHost(current.host, current.client.activeMonitoring);
	I.validateHost(current.host, false);

});

Data(testdata).Scenario('TS_005 - TC_002 - Edytuj hosta', (I, current) => {

	I.editHost(current.host);
	I.validateHost(current.host, true);

});

Data(testdata).Scenario('TS_005 - TC_003 - Usuń hosta', (I, current) => {

	I.deleteHost(current.host);

});

AfterSuite( (I) => {
	I.amOnPage('/');
	// I.amOnConfigMainPage();
	// I.loginAs('operator');
	// I.deleteClient(client);
});
```

### Details

* CodeceptJS version: v2.0.1
* NodeJS Version: from docker codeception/codeceptjs
* Operating System: from docker codeception/codeceptjs
* Protractor: from docker codeception/codeceptjs

**Comment by DavertMik** [comment]
Please don't use `I` inside BeforeSuite.
You call to `I` before a test starts, so Helpers may not be initialized yet.

**Comment by AdamOakman** [comment]
@DavertMik thank you for the answer

Any suggestions, workarounds how to achieve what I'm trying to achieve? I want to have a complex setup using protractor before actual test start. How to use hooks so I can perform some selenium actions **only once** before and after all tests inside single suite?

**Comment by xMutaGenx** [comment]
In theory best practice probably would be to have clean browser with restarts and treat every test with it's own before/after , however if you **really** need selenium actions before the whole suite of tests you can do something like:

```
Feature('Site Loads');

let requireSetup = true;

Before( (I) => {
    if(requireSetup){
        I.amOnPage('/');
        requireSetup = false;
    }
});
```

I would not recommend that however. Hope that helps.

**Comment by AdamOakman** [comment]
Actually it is a _clean code_ request. In my case, I can just write Scenario "TC_000" for "before all" operations and "TC_004" for "after all" and it would work. However, these operations are not in scope of this particular test suite, so I assume they already work as expected or they were tested successfully in previous suites.

For some reasons other tools like cucumber don't have BeforeAll or BeforeSuite annotations but on the other hand testng supports them and it is possible to run selenium steps inside. I think my proposition would be just elegant way to organize code in a nice understandable shape.
